File: python_backnd/app/routes/services.py
"""
Service routes - CRUD operations for professional services
Like Urban Company / Zomato style service listings
"""

from fastapi import APIRouter, HTTPException, status, Depends
from typing import List, Optional
from datetime import datetime
from bson import ObjectId

from app.database import get_database, get_users_collection
from app.models.service import ServiceCreate, ServiceUpdate, service_helper
from app.models.user import UserRole
from app.utils.security import get_current_user
from app.middleware.rbac import require_roles


router = APIRouter()


async def get_services_collection():
    """Get services collection from database"""
    db = get_database()
    return db["services"]


async def get_professional_data(professional_id):
    """Get professional user data"""
    users = get_users_collection()
    if isinstance(professional_id, str):
        professional_id = ObjectId(professional_id)
    return await users.find_one({"_id": professional_id})


# ============ CREATE SERVICE ============

@router.post("/", response_model=dict, status_code=status.HTTP_201_CREATED)
async def create_service(
    service_data: ServiceCreate,
    current_user: dict = Depends(get_current_user)
):
    """Create a new service (Professional only)"""
    # Check if user is a professional
    if current_user.get("role") != "professional":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only professionals can create services"
        )
    
    # Check if professional is approved
    if current_user.get("approval_status") != "approved":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Your professional account must be approved to create services"
        )
    
    services = await get_services_collection()
    
    # Get professional details from approval_data
    approval_data = current_user.get("approval_data", {})
    
    # Prepare service document with all new fields
    service_doc = {
        **service_data.model_dump(),
        "professional_id": current_user["_id"],
        "professional_name": current_user.get("name", "Unknown"),
        "professional_image": current_user.get("profile_image"),
        "professional_rating": approval_data.get("rating"),
        "professional_experience": approval_data.get("experience"),
        "professional_verified": current_user.get("is_verified", False),
        "bookings_count": 0,
        "rating": 0.0,
        "total_ratings": 0,
        "reviews_count": 0,
        "earnings": 0.0,
        "created_at": datetime.utcnow(),
        "updated_at": datetime.utcnow()
    }
    
    # Insert service
    result = await services.insert_one(service_doc)
    service_doc["_id"] = result.inserted_id
    
    return {
        "message": "Service created successfully",
        "service": service_helper(service_doc)
    }


# ============ GET MY SERVICES ============

@router.get("/my-services", response_model=dict)
async def get_my_services(
    is_active: Optional[bool] = None,
    category: Optional[str] = None,
    search: Optional[str] = None,
    skip: int = 0,
    limit: int = 100,
    current_user: dict = Depends(get_current_user)
):
    """Get all services for the current professional"""
    if current_user.get("role") != "professional":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only professionals can access their services"
        )
    
    services = await get_services_collection()
    
    # Build query
    query = {"professional_id": current_user["_id"]}
    
    if is_active is not None:
        query["is_active"] = is_active
    
    if category:
        query["category"] = category
    
    if search:
        query["$or"] = [
            {"name": {"$regex": search, "$options": "i"}},
            {"description": {"$regex": search, "$options": "i"}}
        ]
    
    # Get total count
    total = await services.count_documents(query)
    
    # Get services
    cursor = services.find(query).sort("created_at", -1).skip(skip).limit(limit)
    service_list = []
    
    async for service in cursor:
        service_list.append(service_helper(service))
    
    # Calculate stats
    all_services_query = {"professional_id": current_user["_id"]}
    all_services_cursor = services.find(all_services_query)
    
    total_services = 0
    active_services = 0
    total_bookings = 0
    total_earnings = 0.0
    total_rating_sum = 0.0
    total_rating_count = 0
    
    async for service in all_services_cursor:
        total_services += 1
        if service.get("is_active", True):
            active_services += 1
        total_bookings += service.get("bookings_count", 0)
        total_earnings += service.get("earnings", 0.0)
        if service.get("total_ratings", 0) > 0:
            total_rating_sum += service.get("rating", 0.0) * service.get("total_ratings", 0)
            total_rating_count += service.get("total_ratings", 0)
    
    avg_rating = round(total_rating_sum / total_rating_count, 1) if total_rating_count > 0 else 0.0
    
    return {
        "total": total,
        "services": service_list,
        "skip": skip,
        "limit": limit,
        "stats": {
            "total_services": total_services,
            "active_services": active_services,
            "total_bookings": total_bookings,
            "total_earnings": total_earnings,
            "avg_rating": avg_rating
        }
    }


# ============ GET ALL SERVICES (PUBLIC) ============

@router.get("/", response_model=dict)
async def get_all_services(
    category: Optional[str] = None,
    search: Optional[str] = None,
    min_price: Optional[float] = None,
    max_price: Optional[float] = None,
    min_rating: Optional[float] = None,
    professional_id: Optional[str] = None,
    service_area: Optional[str] = None,
    emergency_available: Optional[bool] = None,
    sort_by: Optional[str] = "recommended",  # recommended, price_low, price_high, rating, newest
    skip: int = 0,
    limit: int = 100
):
    """Get all active services (public - for customers) - Urban Company style"""
    services = await get_services_collection()
    users = get_users_collection()
    
    # Build query - only show active services
    query = {"is_active": True}
    
    if category:
        query["category"] = {"$regex": category, "$options": "i"}
    
    if search:
        query["$or"] = [
            {"name": {"$regex": search, "$options": "i"}},
            {"description": {"$regex": search, "$options": "i"}},
            {"category": {"$regex": search, "$options": "i"}},
            {"tags": {"$regex": search, "$options": "i"}}
        ]
    
    if min_price is not None:
        query["price"] = {"$gte": min_price}
    
    if max_price is not None:
        if "price" in query:
            query["price"]["$lte"] = max_price
        else:
            query["price"] = {"$lte": max_price}
    
    if min_rating is not None:
        query["rating"] = {"$gte": min_rating}
    
    if professional_id and ObjectId.is_valid(professional_id):
        query["professional_id"] = ObjectId(professional_id)
    
    if service_area:
        query["service_area"] = {"$regex": service_area, "$options": "i"}
    
    if emergency_available is not None:
        query["emergency_available"] = emergency_available
    
    # Get total count
    total = await services.count_documents(query)
    
    # Sort options
    sort_options = {
        "recommended": [("rating", -1), ("bookings_count", -1)],
        "price_low": [("price", 1)],
        "price_high": [("price", -1)],
        "rating": [("rating", -1), ("total_ratings", -1)],
        "newest": [("created_at", -1)],
        "popular": [("bookings_count", -1)]
    }
    sort_order = sort_options.get(sort_by, sort_options["recommended"])
    
    # Get services
    cursor = services.find(query).sort(sort_order).skip(skip).limit(limit)
    service_list = []
    
    async for service in cursor:
        # Get professional details if not already in service
        professional = None
        if not service.get("professional_image") or not service.get("professional_verified"):
            professional = await get_professional_data(service.get("professional_id"))
        
        service_list.append(service_helper(service, professional))
    
    # Get unique categories for filtering
    categories = await services.distinct("category", {"is_active": True})
    
    return {
        "total": total,
        "services": service_list,
        "skip": skip,
        "limit": limit,
        "categories": sorted(categories),
        "filters_applied": {
            "category": category,
            "search": search,
            "min_price": min_price,
            "max_price": max_price,
            "min_rating": min_rating,
            "service_area": service_area,
            "emergency_available": emergency_available,
            "sort_by": sort_by
        }
    }


# ============ GET SERVICE BY ID ============

@router.get("/{service_id}", response_model=dict)
async def get_service(service_id: str):
    """Get a single service by ID"""
    services = await get_services_collection()
    
    if not ObjectId.is_valid(service_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid service ID"
        )
    
    service = await services.find_one({"_id": ObjectId(service_id)})
    
    if not service:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found"
        )
    
    return {"service": service_helper(service)}


# ============ UPDATE SERVICE ============

@router.put("/{service_id}", response_model=dict)
async def update_service(
    service_id: str,
    service_data: ServiceUpdate,
    current_user: dict = Depends(get_current_user)
):
    """Update a service (Owner only)"""
    services = await get_services_collection()
    
    if not ObjectId.is_valid(service_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid service ID"
        )
    
    # Find existing service
    existing_service = await services.find_one({"_id": ObjectId(service_id)})
    
    if not existing_service:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found"
        )
    
    # Check ownership
    if str(existing_service["professional_id"]) != str(current_user["_id"]):
        # Allow admin to update any service
        if current_user.get("role") not in ["admin", "manager"]:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to update this service"
            )
    
    # Prepare update data
    update_data = service_data.model_dump(exclude_unset=True)
    update_data["updated_at"] = datetime.utcnow()
    
    # Update service
    await services.update_one(
        {"_id": ObjectId(service_id)},
        {"$set": update_data}
    )
    
    # Get updated service
    updated_service = await services.find_one({"_id": ObjectId(service_id)})
    
    return {
        "message": "Service updated successfully",
        "service": service_helper(updated_service)
    }


# ============ TOGGLE SERVICE STATUS ============

@router.patch("/{service_id}/toggle-status", response_model=dict)
async def toggle_service_status(
    service_id: str,
    current_user: dict = Depends(get_current_user)
):
    """Toggle service active/inactive status (Owner only)"""
    services = await get_services_collection()
    
    if not ObjectId.is_valid(service_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid service ID"
        )
    
    # Find existing service
    existing_service = await services.find_one({"_id": ObjectId(service_id)})
    
    if not existing_service:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found"
        )
    
    # Check ownership
    if str(existing_service["professional_id"]) != str(current_user["_id"]):
        if current_user.get("role") not in ["admin", "manager"]:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to update this service"
            )
    
    # Toggle status
    new_status = not existing_service.get("is_active", True)
    
    await services.update_one(
        {"_id": ObjectId(service_id)},
        {"$set": {"is_active": new_status, "updated_at": datetime.utcnow()}}
    )
    
    # Get updated service
    updated_service = await services.find_one({"_id": ObjectId(service_id)})
    
    return {
        "message": f"Service {'activated' if new_status else 'deactivated'} successfully",
        "service": service_helper(updated_service)
    }


# ============ DELETE SERVICE ============

@router.delete("/{service_id}", response_model=dict)
async def delete_service(
    service_id: str,
    current_user: dict = Depends(get_current_user)
):
    """Delete a service (Owner only)"""
    services = await get_services_collection()
    
    if not ObjectId.is_valid(service_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid service ID"
        )
    
    # Find existing service
    existing_service = await services.find_one({"_id": ObjectId(service_id)})
    
    if not existing_service:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found"
        )
    
    # Check ownership
    if str(existing_service["professional_id"]) != str(current_user["_id"]):
        if current_user.get("role") not in ["admin", "manager"]:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to delete this service"
            )
    
    # Delete service
    await services.delete_one({"_id": ObjectId(service_id)})
    
    return {"message": "Service deleted successfully"}


# ============ GET SERVICES BY PROFESSIONAL ============

@router.get("/professional/{professional_id}", response_model=dict)
async def get_services_by_professional(
    professional_id: str,
    skip: int = 0,
    limit: int = 100
):
    """Get all active services for a specific professional (public)"""
    services = await get_services_collection()
    
    if not ObjectId.is_valid(professional_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid professional ID"
        )
    
    # Build query - only show active services for public
    query = {
        "professional_id": ObjectId(professional_id),
        "is_active": True
    }
    
    # Get total count
    total = await services.count_documents(query)
    
    # Get services
    cursor = services.find(query).sort("created_at", -1).skip(skip).limit(limit)
    service_list = []
    
    async for service in cursor:
        service_list.append(service_helper(service))
    
    return {
        "total": total,
        "services": service_list,
        "skip": skip,
        "limit": limit
    }


# ============ GET SERVICE CATEGORIES ============

@router.get("/categories/list", response_model=dict)
async def get_service_categories():
    """Get all unique service categories"""
    services = await get_services_collection()
    
    # Get distinct categories
    categories = await services.distinct("category")
    
    return {
        "categories": sorted(categories)
    }

File: fontend/utils/services.ts
/**
 * Services API service
 * Urban Company / Zomato style service listings
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL?.replace('/api/v1', '/api') || 'http://localhost:8000/api';

// Types
export interface Service {
  id: string;
  name: string;
  category: string;
  description: string;
  price: number;
  price_type: 'fixed' | 'starting_from' | 'hourly';
  duration: string;
  image: string | null;
  gallery_images: string[];
  is_active: boolean;
  
  // Service Details
  service_area: string | null;
  service_includes: string[];
  service_excludes: string[];
  tags: string[];
  
  // Service Features
  features: string[];
  warranty_days: number | null;
  
  // Availability
  available_days: string[];
  available_time_start: string | null;
  available_time_end: string | null;
  emergency_available: boolean;
  
  // Professional Info
  professional_id: string;
  professional_name: string;
  professional_image: string | null;
  professional_rating: number | null;
  professional_experience: string | null;
  professional_verified: boolean;
  
  // Stats
  bookings_count: number;
  rating: number;
  total_ratings: number;
  reviews_count: number;
  earnings: number;
  created_at: string;
  updated_at: string;
}

export interface ServiceCreate {
  name: string;
  category: string;
  description: string;
  price: number;
  price_type?: string;
  duration: string;
  image?: string;
  gallery_images?: string[];
  is_active?: boolean;
  
  // Service Details
  service_area?: string;
  service_includes?: string[];
  service_excludes?: string[];
  tags?: string[];
  
  // Service Features
  features?: string[];
  warranty_days?: number;
  
  // Availability
  available_days?: string[];
  available_time_start?: string;
  available_time_end?: string;
  emergency_available?: boolean;
}

export interface ServiceUpdate {
  name?: string;
  category?: string;
  description?: string;
  price?: number;
  price_type?: string;
  duration?: string;
  image?: string;
  gallery_images?: string[];
  is_active?: boolean;
  
  // Service Details
  service_area?: string;
  service_includes?: string[];
  service_excludes?: string[];
  tags?: string[];
  
  // Service Features
  features?: string[];
  warranty_days?: number;
  
  // Availability
  available_days?: string[];
  available_time_start?: string;
  available_time_end?: string;
  emergency_available?: boolean;
}

export interface ServiceStats {
  total_services: number;
  active_services: number;
  total_bookings: number;
  total_earnings: number;
  avg_rating: number;
}

export interface ServicesResponse {
  total: number;
  services: Service[];
  skip: number;
  limit: number;
  stats?: ServiceStats;
  categories?: string[];
  filters_applied?: {
    category: string | null;
    search: string | null;
    min_price: number | null;
    max_price: number | null;
    min_rating: number | null;
    service_area: string | null;
    emergency_available: boolean | null;
    sort_by: string;
  };
}

export interface ServiceResponse {
  message: string;
  service: Service;
}

// Professional Profile Types
export interface ProfessionalProfile {
  id: string;
  name: string;
  email: string;
  phone: string;
  profile_image: string | null;
  is_verified: boolean;
  approval_status: string;
  
  // Business Profile
  bio: string;
  experience: string;
  hourly_rate: number | null;
  service_areas: string[];
  skills: string[];
  languages: string[];
  certifications: string[];
  working_hours_start: string;
  working_hours_end: string;
  working_days: string[];
  emergency_available: boolean;
  
  // Address
  address: string;
  city: string;
  state: string;
  pincode: string;
  
  // Professional category
  category: string;
  sub_category: string;
  
  // Stats
  rating: number;
  total_reviews: number;
  total_bookings: number;
  member_since: string | null;
}

export interface ProfessionalPublicProfile {
  id: string;
  name: string;
  phone: string | null;
  profile_image: string | null;
  is_verified: boolean;
  approval_status: string;
  bio: string;
  experience: string;
  hourly_rate: number | null;
  category: string;
  sub_category: string;
  service_areas: string[];
  skills: string[];
  languages: string[];
  certifications: string[];
  working_hours_start: string;
  working_hours_end: string;
  working_days: string[];
  emergency_available: boolean;
  city: string;
  state: string;
  rating: number;
  total_reviews: number;
  total_bookings: number;
  member_since: string | null;
}

// API Error class
export class APIError extends Error {
  status: number;
  
  constructor(message: string, status: number) {
    super(message);
    this.status = status;
    this.name = 'APIError';
  }
}

// Get access token
function getAccessToken(): string | null {
  if (typeof window === 'undefined') return null;
  return localStorage.getItem('access_token');
}

// Helper function for API requests
async function apiRequest<T>(
  endpoint: string,
  options: RequestInit = {}
): Promise<T> {
  const url = `${API_BASE_URL}${endpoint}`;
  
  const headers: HeadersInit = {
    'Content-Type': 'application/json',
    ...options.headers,
  };

  // Add auth token if available
  const token = getAccessToken();
  if (token) {
    (headers as Record<string, string>)['Authorization'] = `Bearer ${token}`;
  }

  const response = await fetch(url, {
    ...options,
    headers,
  });

  const data = await response.json();

  if (!response.ok) {
    console.error('API Error Response:', {
      status: response.status,
      statusText: response.statusText,
      data: data
    });
    
    if (response.status === 422 && data.detail) {
      const errorMessage = Array.isArray(data.detail)
        ? data.detail.map((err: any) => `${err.loc.join('.')}: ${err.msg}`).join(', ')
        : data.detail;
      throw new APIError(errorMessage, response.status);
    }
    
    throw new APIError(data.detail || 'An error occurred', response.status);
  }

  return data;
}

// ============ SERVICE API FUNCTIONS ============

/**
 * Create a new service
 */
export async function createService(serviceData: ServiceCreate): Promise<ServiceResponse> {
  return apiRequest<ServiceResponse>('/services/', {
    method: 'POST',
    body: JSON.stringify(serviceData),
  });
}

/**
 * Get my services (for professionals)
 */
export async function getMyServices(params?: {
  is_active?: boolean;
  category?: string;
  search?: string;
  skip?: number;
  limit?: number;
}): Promise<ServicesResponse> {
  const searchParams = new URLSearchParams();
  
  if (params?.is_active !== undefined) {
    searchParams.append('is_active', String(params.is_active));
  }
  if (params?.category) {
    searchParams.append('category', params.category);
  }
  if (params?.search) {
    searchParams.append('search', params.search);
  }
  if (params?.skip !== undefined) {
    searchParams.append('skip', String(params.skip));
  }
  if (params?.limit !== undefined) {
    searchParams.append('limit', String(params.limit));
  }
  
  const queryString = searchParams.toString();
  const endpoint = `/services/my-services${queryString ? `?${queryString}` : ''}`;
  
  return apiRequest<ServicesResponse>(endpoint);
}

/**
 * Get all services (public) - Urban Company style with advanced filters
 */
export async function getAllServices(params?: {
  category?: string;
  search?: string;
  min_price?: number;
  max_price?: number;
  min_rating?: number;
  service_area?: string;
  emergency_available?: boolean;
  sort_by?: 'recommended' | 'price_low' | 'price_high' | 'rating' | 'newest' | 'popular';
  skip?: number;
  limit?: number;
}): Promise<ServicesResponse> {
  const searchParams = new URLSearchParams();
  
  if (params?.category) {
    searchParams.append('category', params.category);
  }
  if (params?.search) {
    searchParams.append('search', params.search);
  }
  if (params?.min_price !== undefined) {
    searchParams.append('min_price', String(params.min_price));
  }
  if (params?.max_price !== undefined) {
    searchParams.append('max_price', String(params.max_price));
  }
  if (params?.min_rating !== undefined) {
    searchParams.append('min_rating', String(params.min_rating));
  }
  if (params?.service_area) {
    searchParams.append('service_area', params.service_area);
  }
  if (params?.emergency_available !== undefined) {
    searchParams.append('emergency_available', String(params.emergency_available));
  }
  if (params?.sort_by) {
    searchParams.append('sort_by', params.sort_by);
  }
  if (params?.skip !== undefined) {
    searchParams.append('skip', String(params.skip));
  }
  if (params?.limit !== undefined) {
    searchParams.append('limit', String(params.limit));
  }
  
  const queryString = searchParams.toString();
  const endpoint = `/services/${queryString ? `?${queryString}` : ''}`;
  
  return apiRequest<ServicesResponse>(endpoint);
}

/**
 * Get service by ID
 */
export async function getServiceById(serviceId: string): Promise<{ service: Service }> {
  return apiRequest<{ service: Service }>(`/services/${serviceId}`);
}

/**
 * Update a service
 */
export async function updateService(serviceId: string, serviceData: ServiceUpdate): Promise<ServiceResponse> {
  return apiRequest<ServiceResponse>(`/services/${serviceId}`, {
    method: 'PUT',
    body: JSON.stringify(serviceData),
  });
}

/**
 * Toggle service status (active/inactive)
 */
export async function toggleServiceStatus(serviceId: string): Promise<ServiceResponse> {
  return apiRequest<ServiceResponse>(`/services/${serviceId}/toggle-status`, {
    method: 'PATCH',
  });
}

/**
 * Delete a service
 */
export async function deleteService(serviceId: string): Promise<{ message: string }> {
  return apiRequest<{ message: string }>(`/services/${serviceId}`, {
    method: 'DELETE',
  });
}

/**
 * Get services by professional ID (public)
 */
export async function getServicesByProfessional(professionalId: string, params?: {
  skip?: number;
  limit?: number;
}): Promise<ServicesResponse> {
  const searchParams = new URLSearchParams();
  
  if (params?.skip !== undefined) {
    searchParams.append('skip', String(params.skip));
  }
  if (params?.limit !== undefined) {
    searchParams.append('limit', String(params.limit));
  }
  
  const queryString = searchParams.toString();
  const endpoint = `/services/professional/${professionalId}${queryString ? `?${queryString}` : ''}`;
  
  return apiRequest<ServicesResponse>(endpoint);
}

/**
 * Get all service categories
 */
export async function getServiceCategories(): Promise<{ categories: string[] }> {
  return apiRequest<{ categories: string[] }>('/services/categories/list');
}

// ============ PROFESSIONAL PROFILE API FUNCTIONS ============

/**
 * Get professional business profile
 */
export async function getProfessionalProfile(): Promise<{ profile: ProfessionalProfile }> {
  return apiRequest<{ profile: ProfessionalProfile }>('/users/professional-profile');
}

/**
 * Update professional business profile
 */
export async function updateProfessionalProfile(data: Partial<{
  bio: string;
  experience: string;
  hourly_rate: number;
  service_areas: string[];
  skills: string[];
  languages: string[];
  certifications: string[];
  working_hours_start: string;
  working_hours_end: string;
  working_days: string[];
  emergency_available: boolean;
  address: string;
  city: string;
  state: string;
  pincode: string;
}>): Promise<{ message: string; user: any }> {
  const params = new URLSearchParams();
  
  Object.entries(data).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      if (Array.isArray(value)) {
        value.forEach(v => params.append(key, v));
      } else {
        params.append(key, String(value));
      }
    }
  });
  
  return apiRequest<{ message: string; user: any }>(`/users/professional-profile?${params.toString()}`, {
    method: 'PUT',
  });
}

/**
 * Get public professional profile
 */
export async function getProfessionalPublicProfile(professionalId: string): Promise<{ professional: ProfessionalPublicProfile }> {
  return apiRequest<{ professional: ProfessionalPublicProfile }>(`/users/professional/${professionalId}/public`);
}
